File: ios/SsdidWallet/Domain/Transport/DriveApi.swift
import Foundation

/// Drive API client for SSDID Drive registration, verification, and authentication.
final class DriveApi {

    private let client: SsdidHttpClient
    private let baseURL: String

    init(client: SsdidHttpClient, baseURL: String) {
        self.client = client
        self.baseURL = baseURL
    }

    // MARK: - Registration

    func register(request: RegisterStartRequest) async throws -> RegisterStartResponse {
        return try await client.post(
            url: "\(baseURL)/api/auth/ssdid/register",
            body: request,
            responseType: RegisterStartResponse.self
        )
    }

    func registerVerify(request: RegisterVerifyRequest) async throws -> RegisterVerifyResponse {
        return try await client.post(
            url: "\(baseURL)/api/auth/ssdid/register/verify",
            body: request,
            responseType: RegisterVerifyResponse.self
        )
    }

    // MARK: - Authentication

    func authenticate(request: DriveAuthenticateRequest) async throws -> DriveAuthenticateResponse {
        return try await client.post(
            url: "\(baseURL)/api/auth/ssdid/authenticate",
            body: request,
            responseType: DriveAuthenticateResponse.self
        )
    }

    // MARK: - Invitations

    func getInvitationByToken(_ token: String) async throws -> InvitationDetailsResponse {
        return try await client.get(
            url: "\(baseURL)/api/invitations/token/\(token)",
            responseType: InvitationDetailsResponse.self
        )
    }

    func acceptWithWallet(token: String, request: AcceptWithWalletRequest) async throws -> AcceptWithWalletResponse {
        return try await client.post(
            url: "\(baseURL)/api/invitations/token/\(token)/accept-with-wallet",
            body: request,
            responseType: AcceptWithWalletResponse.self
        )
    }
}

File: android/sdk/ssdid-core/src/main/java/my/ssdid/sdk/domain/transport/DriveApi.kt
package my.ssdid.sdk.domain.transport

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import my.ssdid.sdk.domain.model.VerifiableCredential
import my.ssdid.sdk.domain.transport.dto.RegisterStartRequest
import my.ssdid.sdk.domain.transport.dto.RegisterStartResponse
import my.ssdid.sdk.domain.transport.dto.RegisterVerifyRequest
import my.ssdid.sdk.domain.transport.dto.RegisterVerifyResponse
import retrofit2.http.Body
import retrofit2.http.POST

interface DriveApi {
    @POST("api/auth/ssdid/register")
    suspend fun register(@Body request: RegisterStartRequest): RegisterStartResponse

    @POST("api/auth/ssdid/register/verify")
    suspend fun registerVerify(@Body request: RegisterVerifyRequest): RegisterVerifyResponse

    @POST("api/auth/ssdid/authenticate")
    suspend fun authenticate(@Body request: DriveAuthenticateRequest): DriveAuthenticateResponse
}

@Serializable
data class DriveAuthenticateRequest(
    val credential: VerifiableCredential,
    @SerialName("challenge_id") val challengeId: String? = null
)

@Serializable
data class DriveAuthenticateResponse(
    @SerialName("session_token") val sessionToken: String,
    val did: String? = null,
    @SerialName("server_did") val serverDid: String,
    @SerialName("server_signature") val serverSignature: String? = null
)
